server/util/rexec/rexec.go

```diff
@@ -0,0 +1,452 @@
+// package rexec provides a command interface similar to the "os/exec" package
+// but runs commands using a remote execution client.
+//
+// Basic usage: you can just call Result() to run a command end-to-end,
+// including uploading the Action, Command, and input root dir to cache:
+//
+//	cmd := rexec.Command(env, "echo", "hello")
+//	res, err := cmd.Result(ctx)
+//	if err != nil {
+//		return err
+//	}
+//	fmt.Println(string(res.Stdout)) // prints "hello"
+//
+// You can also manually walk the command through its lifecycle:
+//
+//	// TODO: handle errors!
+//	cmd := rexec.Command(env, "echo" "hello")
+//	err := cmd.Upload(ctx)
+//	err = cmd.Start(ctx)
+//	err = cmd.Wait(ctx)
+//	res, err := cmd.Result(ctx)
+//
+// Command properties are configured by setting attributes on *Cmd:
+//
+//	cmd.Dir = "/path/to/input_root_dir"
+//	cmd.Env = []string{"FOO=BAR"}
+//	cmd.Platform = []string{"recycle-runner=true"}
```

Comment:
I'm kinda meh on this bit -- maybe it's the java in me but I feel like classes should be responsible for their own state and nobody else should be reaching in and touching stuff like this. I think it'd be clearer if you disallowed this and added an option to provide a CommandProto if the caller wants to override it, otherwise you build your own.

Reply:
I went ahead and made this refactor since it feels much cleaner this way, and later if we do want a higher-level / idiomatic construct like `*Cmd` we can build something like that using these lower level primitives.